File: _src/main/java/raccoonlang/parser/ast/FqtnAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FqtnAstNode {

    public List<Token> identifiers = new ArrayList<>();
    public GenericTypesAstNode genericTypesAstNode;

    public static FqtnAstNode parse(Parser parser) {
        FqtnAstNode fqtnAstNode = new FqtnAstNode();

        fqtnAstNode.identifiers.add(parser.take(TokenType.IDENTIFIER));

        while(parser.peek().getTokenType() == TokenType.DOT){
            parser.take();
            fqtnAstNode.identifiers.add(parser.take(TokenType.IDENTIFIER));
        }

        fqtnAstNode.genericTypesAstNode = GenericTypesAstNode.tryParse(parser);

        return fqtnAstNode;
    }

    @Override
    public String toString() {
        return "FqtnAstNode{" +
                "identifiers=" + identifiers.stream().map(x -> x.getText()).reduce((a, b) -> a + "." + b).orElse("") +
                '}';
    }
}

File: Parser/AST/FqtnAstNode.cs
namespace Raccoonlang.Parser.AST;

using Tokenizer;

// fully qualified type name
public class FqtnAstNode 
{
    public List<Token> Identifiers = new List<Token>();
    public GenericTypesAstNode? Node;

    public static FqtnAstNode Parse(Parser parser)
    {
        FqtnAstNode node = new FqtnAstNode();

        node.Identifiers.Add(parser.Take(TokenType.IDENTIFIER));

        while(parser.Peek().Type == TokenType.DOT) {
            parser.Take();
            node.Identifiers.Add(parser.Take(TokenType.IDENTIFIER));
        }

        node.Node = GenericTypesAstNode.TryParse(parser);
        if (node.Node == null) throw new NullReferenceException("Node is null!");
        return node;
    }

    public override string ToString() {
        return "FqtnAstNode{identifiers=" + Identifiers.ToArray() + "}";
    }
}